### PR TITLE
chore(tracer): deprecate Tracer.writer

### PR DESCRIPTION
**Pull request by majorgreys**
The `Tracer.writer` will be removed from the public API in v1.0 so is deprecated here.

### CODE REVIEW

File: ddtrace/tracer.py [hunk 1]
```diff
@@ -811,7 +812,16 @@ def write(self, spans):
...
-            self.writer.write(spans=spans)
+            self._writer.write(spans=spans)
+
+    @removals.removed_property(message="Use Tracer.flush instead to flush buffered traces to agent", version="1.0.0")
+    def writer(self):
+        return self._writer
+
+    def flush(self):
+        """Flush buffered traces."""
+        if isinstance(self._writer, AgentWriter):
```
Comment (brettlangdon) [inline ddtrace/tracer.py:830]: Should the base writer have a flush method that is a no-op by default?

File: releasenotes/notes/deprecate-tracer-writer-b4921987ad47a832.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+deprecations:
+  - |
+    :py:attr:`Tracer.writer` is deprecated. To force flushing of buffered traces to the agent, use :py:meth:`Tracer.flush` instead.
```
Comment (brettlangdon) [inline releasenotes/notes/deprecate-tracer-writer-b4921987ad47a832.yaml:4]: does this linking work as expected? do you need to include the module namespace for it as well? `ddtrace.tracer.Tracer.writer` ?

File: ddtrace/tracer.py (hunk 1 above)
Comment (majorgreys) [reply ddtrace/tracer.py:832]: I was reluctant here to call the method `flush` but if we take a page from `IOBase` where `flush` is a no-op for non-blocking streams. Will make this internal module change as it is not adding any new functionality really, just a refactor.

**Review by mabdinur — APPROVED** [review]
LGTM

File: releasenotes/notes/deprecate-tracer-writer-b4921987ad47a832.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+deprecations:
+  - |
+    :py:attr:`ddtrace.Tracer.writer` is deprecated. To force flushing of buffered traces to the agent, use :py:meth:`ddtrace.Tracer.flush` instead.
```
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply releasenotes/notes/deprecate-tracer-writer-b4921987ad47a832.yaml:4]: Also should we have separate feature entries for the new APIs?

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -811,7 +812,24 @@ def write(self, spans):
...
+        # type: () -> Optional[str]
+        """Trace agent url"""
+        if isinstance(self._writer, AgentWriter):
+            return self._writer.agent_url
+
+        return None
+
+    def flush(self):
+        """Flush the buffer of the trace writer. This does nothing if an unbuffered trace writer is used."""
+        self._writer.flush_queue()
```
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/tracer.py:832]: This was added in this PR:

https://github.com/DataDog/dd-trace-py/pull/3223/files#diff-0b32c6cdae8bdd01786317bcabebf73dbb57e535adab74c94858fda05da411efR176

**Review by mabdinur — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**